pyccel/parser/semantic.py

```diff
@@ -2644,6 +2646,10 @@ def _visit_DottedName(self, expr):
 
             args = [FunctionCallArgument(visited_lhs), *self._handle_function_args(rhs.args)]
             method = cls_base.get_method(rhs_name)
+            if isinstance(method, PyccelFunctionDef):
+                annotation_method = '_visit_' + method.cls_name.__name__
+                if hasattr(self, annotation_method):
+                    return getattr(self, annotation_method)(visited_lhs, rhs.args[0].value)
```

Comment:
This `_visit` syntax does not respect the docs:
https://github.com/pyccel/pyccel/blob/devel/developer_docs/semantic_stage.md#navigation

Specifically:
> These _visit_X functions must have the form:
> ```python
> def _visit_ClassName(self, stmt):
>     ...
> ```
It does not seem to be mentioned in the documentation but the logic of a `_visit(` function is always the same. It takes a syntactic object and returns a semantic object. This rule is also not respected here. I will create a PR to clarify this point in the docs.

For this PR there are 2 solutions.
1. Rename the `_visit_ListExtend` to something like `_handle_ListExtend` this makes it clear that this function does not follow the standard rules but this will require you to write a docstring for it and I don't think the code above makes sense in this case as the function call will be specific to each function being implemented.
2. Modify this code to match the conventions.

I prefer the second option for several reasons:
1. You can move the call up a few lines to avoid visiting the `args` and then discarding the result
2. In the function `_visit_ListExtend` you need to work with syntactic objects so it is easier to collect a syntactic object and unpack what you need rather than trying to guess what you need
3. The same prototype can be used no matter the number of method arguments
4. We are not mixing syntactic and semantic objects (here the first argument is semantic but the second is syntactic
5.  It is easier to get an ast node to raise any errors that are needed. `expr.python_ast` exists and is well defined for a `DottedName` but it does not exist for something like a `PyccelSymbol` as the same symbol is used in lots of places. Here you can collect a `python_ast` from `expr` (but it is not passed), from `rhs.args[0]` (but it is not passed), from `visited_lhs` if it is a non-trivial expression or from `rhs.args[0].value` if it is a non-trivial expression. However `a.extend(b)` is a common use-case where you will only pass trivial expressions which don't provide the `python_ast`

Reply:
can you check 6ce1835 and give me your feedback if possible please

Reply:
This looks much better :+1: I just have one comment (I'll leave a comment on the code)